openmsupply/remote-server
# 371 add lines to insert supplier invoice mutation

Fixes: #371 

Expands on downstream branches, main difference is errors in invoice lines are an array and are nested in 'lines' key of InsertSupplierInvoiceError. Also simplified full invoice shape (which will change again in next PR, as we would need to do delete and or insert and or update of invoice lines in one transaction). I tried to make errors a bit more generic in graphql, which ended up really hard to follow (it's already getting a bit complex), so ended up not committing those changes.

Also NaiveDate seems to work very good (i think we should use inbuilt async graphql functionality for this), had to enable chrono feature.

### Testing

* After full test with postgres, point base.yaml to `omsupply-database-gql-invoices-query`
* Run with postgres features (remove sync in main)
* Add at least on name_store_join with name_is_supplier `insert into name_store_join (id, name_is_supplier, name_is_customer, store_id, name_id) values('a', true, false, 'store_a', 'name_store_a')`

Use query below (try changing things for errors to show up, check `InsertSupplierInvoiceLineError` in graphql introspection or code)

```graphql
mutation mutInsert {
  insertSupplierInvoice(
    input: { id: "ABCD", otherPartyId: "name_store_a", status: "CONFIRMED", lines: [{
      id: "ABC2",
      itemId: "item_b",
      packSize: 1,
      costPricePerPack: 2,
      sellPricePerPack: 3,
      expiryDate: "2012-11-12",
      numberOfPacks: 10
      
    },{
      id: "B3",
      itemId: "item_a",
      packSize: 1,
      costPricePerPack: 2,
      sellPricePerPack: 3,
      expiryDate: "2012-11-12",
      numberOfPacks: 3
      
    }] }
  ) {
    ... on Invoice {
      id
    }

    ... on InsertSupplierInvoiceErrors {
      id
      errors {
        ... on InsertSupplierInvoiceError {
          __typename
          description
        }
        ... on ForeignKeyError {
          __typename
          key
          id
        }
        ... on OtherPartyNotASuppier {
          __typename
          otherParty {
            id
            name
            isSupplier
            isCustomer
          }
        }
        ... on Dberror {
          fullError
        }
      }
      lines {
        ... on InsertSupplierInvoiceLineErrors {
           __typename
          id
          errors {
            ... on InsertSupplierInvoiceError {
              __typename
              description
            }
            ... on ValueOutOfRange {
              field
            }
            ... on ForeignKeyError {
              key
              id
            }
            ... on RecordAlreadyExist  {
              description
            }
          }
        }
      }
    }
  }
}
```
